File: Fbs.WebApi/Repository/FacilityRepository.cs
using System.Linq.Expressions;
using Fbs.WebApi.Entities;
using Fbs.WebApi.Options;
using Google.Apis.Sheets.v4;
using Microsoft.Extensions.Caching.Hybrid;
using Microsoft.Extensions.Options;

namespace Fbs.WebApi.Repository;

public class FacilityRepository(
    InstrumentationSource instrumentation,
    HybridCache cache,
    IOptions<GoogleOptions> options,
    SheetsService sheetsService
) : IRepository<Facility>
{
    private readonly string[] _header =
    [
        "Name",
        "Group",
        "Scope"
    ];

    public async Task<List<Facility>> GetListAsync(CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await cache.GetOrCreateAsync(
            "Facilities",
            (sheetsService, options),
            async (state, ct) =>
                await state.sheetsService.Spreadsheets.Values.Get(state.options.Value.SpreadsheetId, "Facilities")
                    .ExecuteAsync(ct),
            cancellationToken: cancellationToken);

        if (!items.Values.First().SequenceEqual(_header))
        {
            throw new Exception("Unexpected headers in Facilities table");
        }

        return items.Values
            .Skip(1)
            .Select((row, idx) => new Facility
            {
                Row = idx + 2,
                Name = row.ElementAtOrDefault(0) as string,
                Group = row.ElementAtOrDefault(1) as string,
                Scope = (row.ElementAtOrDefault(2) as string)?
                    .Split(",")
                    .Select(s => s.Trim())
                    .Where(s => !string.IsNullOrEmpty(s))
                    .ToList(),
            })
            .ToList();
    }

    public async Task<Facility?> FindAsync(Expression<Func<Facility, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.SingleOrDefault(predicate.Compile());
    }

    public async Task<Facility> GetAsync(Expression<Func<Facility, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var items = await GetListAsync(cancellationToken);
        return items.Single(predicate.Compile());
    }

    public async Task<Facility> InsertAsync(Facility entity, CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }

    public async Task<Facility> UpdateAsync(Facility entity, CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }

    public async Task DeleteAsync(Expression<Func<Facility, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }
}

File: Fbs.WebApi.Kotlin/src/main/kotlin/sg/from/fbs/webapi/repository/FacilityRepository.kt
package sg.from.fbs.webapi.repository

import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Repository
import sg.from.fbs.webapi.entity.FacilityEntity
import sg.from.fbs.webapi.integration.GoogleSheetsClient

@Repository
class FacilityRepository(private val sheets: GoogleSheetsClient) {
    private val header = listOf("Name", "Group", "Scope")

    @Cacheable("Facilities")
    fun getList(): List<FacilityEntity> {
        val rows = sheets.readValues("Facilities")
        require(rows.firstOrNull() == header) { "Unexpected headers in Facilities table" }
        return rows.drop(1).mapIndexed { idx, row ->
            FacilityEntity(
                row = idx + 2,
                name = row.getOrNull(0),
                group = row.getOrNull(1),
                scope = row.getOrNull(2)?.split(",")?.map { it.trim() }?.filter { it.isNotBlank() },
            )
        }
    }

    fun findByName(name: String): FacilityEntity? = getList().singleOrNull { it.name == name }

    fun getByName(name: String): FacilityEntity = getList().single { it.name == name }
}
